core/services/keystore/models.go

```diff
@@ -31,50 +32,103 @@ type encryptedKeyRing struct {
 	EncryptedKeys []byte
 }
 
-func (ekr encryptedKeyRing) Decrypt(password string) (keyRing, error) {
+func (ekr encryptedKeyRing) Decrypt(password string) (*keyRing, error) {
 	if len(ekr.EncryptedKeys) == 0 {
 		return newKeyRing(), nil
 	}
 	var cryptoJSON gethkeystore.CryptoJSON
 	err := json.Unmarshal(ekr.EncryptedKeys, &cryptoJSON)
 	if err != nil {
-		return keyRing{}, err
+		return nil, err
 	}
 	marshalledRawKeyRingJson, err := gethkeystore.DecryptDataV3(cryptoJSON, adulteratedPassword(password))
 	if err != nil {
-		return keyRing{}, err
+		return nil, err
 	}
 	var rawKeys rawKeyRing
 	err = json.Unmarshal(marshalledRawKeyRingJson, &rawKeys)
 	if err != nil {
-		return keyRing{}, err
+		return nil, err
 	}
 	ring, err := rawKeys.keys()
 	if err != nil {
-		return keyRing{}, err
+		return nil, err
 	}
 	return ring, nil
 }
 
 type keyStates struct {
-	Eth map[string]*ethkey.State
+	// Key ID => chain ID => state
+	KeyIDChainID map[string]map[string]*ethkey.State
+	// Chain ID => Key ID => state
+	ChainIDKeyID map[string]map[string]*ethkey.State
+	All          []*ethkey.State
 }
 
-func newKeyStates() keyStates {
-	return keyStates{
-		Eth: make(map[string]*ethkey.State),
+func newKeyStates() *keyStates {
+	return &keyStates{
+		KeyIDChainID: make(map[string]map[string]*ethkey.State),
+		ChainIDKeyID: make(map[string]map[string]*ethkey.State),
 	}
 }
 
-func (ks keyStates) validate(kr keyRing) (err error) {
-	for id := range kr.Eth {
-		_, exists := ks.Eth[id]
-		if !exists {
-			err = multierr.Combine(err, errors.Errorf("key %s is missing state", id))
-		}
+// warning: not thread-safe! caller must sync
+// adds or replaces a state
+func (ks *keyStates) add(state *ethkey.State) {
+	cid := state.EVMChainID.String()
+	kid := state.KeyID()
+
+	keyStates, exists := ks.KeyIDChainID[kid]
+	if !exists {
+		keyStates = make(map[string]*ethkey.State)
+		ks.KeyIDChainID[kid] = keyStates
+	}
+	keyStates[cid] = state
+
+	chainStates, exists := ks.ChainIDKeyID[cid]
+	if !exists {
+		chainStates = make(map[string]*ethkey.State)
+		ks.ChainIDKeyID[cid] = chainStates
+	}
+	chainStates[kid] = state
+
+	ks.All = append(ks.All, state)
+}
+
+// warning: not thread-safe! caller must sync
+func (ks *keyStates) get(addr common.Address, chainID *big.Int) *ethkey.State {
+	chainStates, exists := ks.KeyIDChainID[addr.Hex()]
+	if !exists {
+		return nil
 	}
+	return chainStates[chainID.String()]
+}
+
+// warning: not thread-safe! caller must sync
+func (ks keyStates) enable(addr common.Address, chainID *big.Int) {
+	state := ks.get(addr, chainID)
+	state.Disabled = false
+}
+
+// warning: not thread-safe! caller must sync
+func (ks keyStates) disable(addr common.Address, chainID *big.Int) {
+	state := ks.get(addr, chainID)
+	state.Disabled = true
+}
 
-	return err
+// warning: not thread-safe! caller must sync
+func (ks keyStates) delete(addr common.Address) {
```

Comment:
I think this one needs to be a pointer receiver b/c of the append (will break on backing array reallocation):
```suggestion
func (ks *keyStates) delete(addr common.Address) {
```

Reply:
Oh, nvm, it only shrinks here inside of `delete`. I guess not an error then :shrug:

Reply:
Made them all pointer receivers anyway for consistency